### ISSUE TITLE
[Question] Enforce blank lines after closures and new lines before returns

### ISSUE THREAD

**Issue by OwenMelbz** [squizlabs/PHP_CodeSniffer#1555]
Hi guys,

Sorry that this is not an actual bug, but are there any pre-existing sniffs or configurations that allows for the following (or know how to add it in)

# New line after closures e.g

```php
array_filter($someArray, function () {
    return true;
});

$cake = 'yes please';
```

to prevent

```php
array_filter($someArray, function () {
    return true;
});
$cake = 'yes please';
```
# Force blank line before return statements (that are not the only statements in the method)

```php
getSomething()
{
    $something = 'something';

    return $something;
}
```

or

```php
getSomething()
{
    return 'something';
}
```

rather than

```php
getSomething()
{
    $something = 'something';
    return $something;
}
```

Thanks :)

**Comment by gsherwood** [comment]
No, there are no included sniffs for either of those rules, and I don't know of any coding standards or custom sniffs that enforce them.

They look pretty specific, so you'd probably need to code up your own sniffs to enforce these rules, unless you know of a publishing coding standard that also uses them and uses PHPCS as well.

They don't actually look that hard to write given you have very specific tokens you can listen for.

In the first case, you can listen for T_CLOSURE, ask PHPCS for the end of the closure, then check how many blank lines are between it and the next bit of PHP code. The only problem I see here is that it really isn't whitespace after a closure. You're asking for a blank line after a function call that contains a closure as an argument. But what if the closure argument is followed by more arguments? And what if the closure is assigned to a variable instead?

For the second, you'd listen for T_RETURN tokens and make sure there is a blank line between it and the previous content. Again, you need to consider other places you use return, like what happens if the return is inside an IF statement (still want a blank line there)? Or what happens if the return is actually the breaking statement of a CASE block (do you have other rules for these)?

I'd offer to help you write these sniffs, but I'm so far behind on PHPCS work that I really need to focus on the PRs and bug fixes first. But if you want to give them a go, I'd be happy to look over the sniffs and provide any help I can.
